### ISSUE TITLE
Refactor training graph generators

### ISSUE THREAD

**Issue by rdadolf** [rdadolf/fathom#15]
The ops required for training (loss function, optimizer, and training loop) are spread out over too many files. They are difficult to read and follow.

These should be integrated into the models themselves, even if it means a little bit of duplication across files. Like the specification of the inference graphs (see #9), a reader who knows the training method behind a given model should be able to recognize it on sight.
